### PR TITLE
Warn if using address overloads on contracts

### LINKED ISSUE #2695: Warn on using address members on contracts
In preparation for #2683.

### PR DESCRIPTION
**Pull request by axic**
Fixes #2695.

### CODE REVIEW

**Comment by axic** [comment]
@chriseth this was an old version I had lying around, perhaps you can rewrite it, at least the tests were done.

**Comment by chriseth** [comment]
Rebased.

**Comment by chriseth** [comment]
@axic please review.

File: libsolidity/analysis/TypeChecker.cpp [hunk 1]
```diff
@@ -1830,6 +1830,16 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
 
 	if (exprType->category() == Type::Category::Contract)
 	{
+		// Warn about using address members on contracts
+		for (auto const& addressMember: IntegerType(160, IntegerType::Modifier::Address).nativeMembers(nullptr))
+			if (addressMember.name == memberName && *annotation.type == *addressMember.type)
```
Comment (axic) [inline libsolidity/analysis/TypeChecker.cpp]: Funny enough, my second commit was (but I didn't submit it with this PR) "Simplify code" and did the same (I think we had a chat to use `nativeMembers`), for some reason though that had 3 conditions:
```
+                       if (member.name == memberName && *member.type == *possibleMembers.front().type && member.declaration == nullptr)
```

I think the last `nullptr` check was just for safety?

File: libsolidity/analysis/TypeChecker.cpp [hunk 2]
```diff
@@ -1830,6 +1830,16 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
...
+				m_errorReporter.warning(
```
Comment (axic) [inline libsolidity/analysis/TypeChecker.cpp]: Actually shouldn't we turn this into an error with 0.5.0?

File: libsolidity/analysis/TypeChecker.cpp (hunk 1 above)
Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp]: Yes, the type is a special type which should not be possible to be used directly.

File: libsolidity/analysis/TypeChecker.cpp (hunk 2 above)
Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp]: We should turn it into an assertion with 0.5.0

File: libsolidity/analysis/TypeChecker.cpp [hunk 3]
```diff
@@ -1831,6 +1831,20 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
 
 	if (exprType->category() == Type::Category::Contract)
 	{
+		// Warn about using address members on contracts
+		bool v050 = m_scope->sourceUnit().annotation().experimentalFeatures.count(ExperimentalFeature::V050);
```
Comment (axic) [inline libsolidity/analysis/TypeChecker.cpp:1835]: `const`? :)

Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp:1835]: Oh come on!

Comment (axic) [reply libsolidity/analysis/TypeChecker.cpp:1835]: I don't care about this, you were putting that on other reviews :)

Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp:1835]: For such cases?

Comment (axic) [reply libsolidity/analysis/TypeChecker.cpp:1835]: Yeah.

**Comment by axic** [comment]
I cannot approve this since its my PR.

**Pull request merged**